chanzuckerberg/idseq-web
# Pipeline auto restart for admin users samples

# Description

Allow auto-restart for admin samples, as long they are using master branch

## Type of change


# Test and Reproduce
Outline the steps to test or reproduce the PR here.

1. Upload a sample as an admin
2. Make it fail due to an unexpected condition
3. If this is the first time it fails for that specific pipeline run id, it should restart
```
PipelineRun
  #update_job_status
    when all stages complete successfully
      changes status to finalized
    when a stage completes with an error
      and it is a known user error
        changes status to failed
        does not send error to airbrake
        does not enqueue new pipeline run
      and it is not a known user error
        and an automatic restart is allowed
          changes status to failed
          sends error to airbrake
          enqueues new pipeline run
        and an automatic restart is not allowed
          changes status to failed
          sends error to airbrake
          does not enqueue new pipeline run
  #automatic_restart_allowed?
    when branch is not master
      should be falsy
    when branch is master
      and sample has no previous pipeline runs with the same pipeline version
        should be truthy
      and sample has previous pipeline runs with the same pipeline version
        and they all succeeded
          should be truthy
        and at least one of them failed
          should be falsy
  #report_failed_pipeline_run_stage
    sends metric to datadog
    when send_to_airbrake is true
      sends error to airbrake and log error
    when send_to_airbrake is false
      do not send error to airbrake and log warn

Finished in 0.66266 seconds (files took 8.98 seconds to load)
17 examples, 0 failures
```
## Impacted Areas in Application
List general components of the application that this PR will affect:

Specific components 

* pipeline_monitor

# Testing Script:

###Single Upload
* Successfully uploaded files from a folder (use test host)
* Pipeline gets initiated successfully 
* Pipeline completes successfully

# Checklist:
